packages/babel-traverse/src/path/family.ts

```diff
@@ -312,6 +312,46 @@ export function getAllPrevSiblings(this: NodePath): NodePath[] {
   return siblings;
 }
 
+type digit = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9";
+
+// convert "1" to 1 (string index to number index)
+type MaybeToIndex<T extends string> = T extends digit ? number : T;
+
+type Pattern<Obj extends string, Prop extends string> = `${Obj}.${Prop}`;
+
+// split "body.body.1" to ["body", "body", 1]
+type Split<P extends string> = P extends Pattern<infer O, infer U>
+  ? [MaybeToIndex<O>, ...Split<U>]
+  : [MaybeToIndex<P>];
+
+// get all K with Node[K] is t.Node | t.Node[]
+type NodeKeyOf<Node extends t.Node | t.Node[]> = keyof Pick<
+  Node,
+  {
+    [Key in keyof Node]-?: Node[Key] extends t.Node | t.Node[] ? Key : never;
+  }[keyof Node]
+>;
+
+// traverse the Node with tuple path ["body", "body", 1]
+// Path should be created with Split
+type Trav<
+  Node extends t.Node | t.Node[],
+  Path extends unknown[],
+> = Path extends [infer K, ...infer R]
+  ? K extends NodeKeyOf<Node>
```

Comment:
Does `Node[K] extends Node | Node[]` work? If it does, it might be better for performance because we don't need to build `NodeKeyOf<Node>`.

Reply:
It doesn't :(

It says `Type 'K' cannot be used to index type 'Node'`, but if I add `K extends keyof Node` before it, it will throw the same error as we ignored.